python/tutorials/11-programmatic-dependent-launch.py

```diff
@@ -0,0 +1,128 @@
+               ):
+    pid = tl.program_id(axis=0)
+    block_start = pid * BLOCK_SIZE
+    offsets = block_start + tl.arange(0, BLOCK_SIZE)
+    mask = offsets < n_elements
+    if USE_GDC:
+        # GDC wait is used to wait for the prior kernel to complete before continuing.
+        # If we utilize Programmatic Dependent Launch, we must wait on the prior kernel
+        # to complete in case a or b are written to by the prior kernel.
+        # This is done to prevent races.
+        tl.extra.cuda.gdc_wait()
+
+    x = tl.load(x_ptr + offsets, mask=mask)
+    y = tl.load(y_ptr + offsets, mask=mask)
+    if USE_GDC:
+        # GDC launch dependents is used to launch dependent kernels.
+        # Once GDC launch it is possible for the next kernel to begin if
+        # there are enough resources.
+        tl.extra.cuda.gdc_launch_dependents()
+    output = x + y
+    tl.store(output_ptr + offsets, output, mask=mask)
+
+
+def add(x: torch.Tensor, y: torch.Tensor, launch_pdl: bool = True):
+    output = torch.empty_like(x)
+    assert x.device == y.device and output.device == x.device
+    n_elements = output.numel()
+    grid = lambda meta: (triton.cdiv(n_elements, meta['BLOCK_SIZE']), )
+    add_kernel[grid](
+        x, y, output, n_elements, BLOCK_SIZE=1024,
+        USE_GDC=launch_pdl,  # set constexpr in kernel to use grid dependence control
+        launch_pdl=launch_pdl,  # launch kernel with PDL flag set enabled
+    )
+    return output
+
+
+def validate(n_elements):
+    x = torch.rand(n_elements, device="cuda", dtype=torch.float32)
+    y = torch.rand(n_elements, device="cuda", dtype=torch.float32)
+
+    torch_result = x + y
+    add_result = add(x, y)
+
+    torch_vs_add = "✅" if torch.allclose(torch_result, add_result, atol=1.0) else "❌"
+    print(f"Number of Elements={n_elements} verification naive vs: ", end="")
+    print(f"add: {torch_vs_add}")
+
+
+@triton.testing.perf_report(
+    triton.testing.Benchmark(
+        x_names=["size"],
+        x_vals=[2**i for i in range(12, 28, 1)],
+        x_log=False,
+        line_arg="provider",
+        line_vals=["pdl-fp32", "fp32"],
+        line_names=["PDL", "No PDL"],
+        styles=[("red", "-"), ("blue", "-")],
+        ylabel='GB/s',
+        plot_name="pdl-performance",
+        args={},
+    ))
+def benchmark(size, provider):
+
+    x = list()
+    y = list()
+
+    for _ in range(100):
+        x.append(torch.rand(size, device="cuda", dtype=torch.float32))
+        y.append(torch.rand(size, device="cuda", dtype=torch.float32))
+
+    start = torch.cuda.Event(enable_timing=True)
+    end = torch.cuda.Event(enable_timing=True)
+
+    g = torch.cuda.CUDAGraph()
+    i = 0
+    with torch.cuda.graph(g):
+        for _ in range(1000):  # 1000 runs
+            add(x[i], y[i], "pdl" in provider)
+            i = (i + 1) % 100
+
+    g.replay()
+    start.record()
+    g.replay()
+    end.record()
+    torch.cuda.synchronize()
+    ms = start.elapsed_time(end) / 1000.0
```

Comment:
Can you use `do_bench_cudagraph` instead? The ratio of boiler plate to relevant code is pretty bad here.

Reply:
I just updated the code to address your concerns. `do_bench_cudagraph` doesn’t loop through multiple buffers, which could cause L2 to cache our tensors and skew the results. To address this, I've now ensured that the function uses large tensors to prevent cross-kernel L2 hits. The size is capped at 3GB, making it compatible with Geforce cards as well.